File: 1851.包含每个查询的最小区间.py
#
# @lc app=leetcode.cn id=1851 lang=python3
#
# [1851] 包含每个查询的最小区间
#

# @lc code=start
from typing import List
import heapq
class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        intervals.sort()
        order = sorted(range(len(queries)), key=lambda x: queries[x])
        pq = []
        ret = [-1] * len(queries)
        idx = 0
        for i in order:
            while idx < len(intervals) and intervals[idx][0] <= queries[i]:
                heapq.heappush(pq, (intervals[idx][1] - intervals[idx][0] + 1, intervals[idx][1]))
                idx += 1
            while pq and pq[0][1] < queries[i]:
                heapq.heappop(pq)
            if pq:
                ret[i] = pq[0][0]
        return ret
# @lc code=end

print(Solution().minInterval(intervals = [[1,4],[2,4],[3,6],[4,4]], queries = [2,3,4,5]))
print(Solution().minInterval(intervals = [[2,3],[2,5],[1,8],[20,25]], queries = [2,19,5,22]))

File: 1851.包含每个查询的最小区间.cpp
/*
 * @lc app=leetcode.cn id=1851 lang=cpp
 *
 * [1851] 包含每个查询的最小区间
 */

// @lc code=start
#include<vector>
#include<queue>
#include<algorithm>
#include<iostream>
using namespace std;
class Solution {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        sort(intervals.begin(), intervals.end());
        vector<int> idxs(queries.size()), ret(queries.size(), -1);
        for(int i=0; i<idxs.size(); ++i) idxs[i] = i;
        sort(idxs.begin(), idxs.end(), [&](int a, int b){return queries[a] < queries[b];});
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        int idx = 0;
        for(auto i: idxs){
            while(idx < intervals.size() && intervals[idx][0] <= queries[i]){
                pq.push({intervals[idx][1] - intervals[idx][0] + 1, intervals[idx][1]});
                ++idx;
            }
            while(!pq.empty() && pq.top().second < queries[i]){
                pq.pop();
            }
            if(!pq.empty()) ret[i] = pq.top().first;
        }
        return ret;
    }
};
// @lc code=end

